### ISSUE TITLE
Apple Silicone Nural Engine: Core ML model package format

### ISSUE THREAD

**Issue by qdrddr** [ollama/ollama#3898]
Please consider adding Core ML model package format support to utilize Apple Silicone Nural Engine + GPU.

List of Core ML package format models

https://github.com/likedan/Awesome-CoreML-Models

**Comment by qdrddr** [comment]
This is about running LLMs locally on Apple Silicone. Core ML is a framework that can redistribute workload across CPU, GPU & Nural Engine (ANE). ANE is available on all modern Apple Devices: iPhones & Macs (A14 or newer and M1 or newer). Ideally, we want to run LLMs on ANE only as it has optimizations for running ML tasks compared to GPU. Apple claims ["deploying your Transformer models on Apple devices with an A14 or newer and M1 or newer chip to achieve up to 10 times faster and 14 times lower peak memory consumption compared to baseline implementations"](https://github.com/apple/ml-ane-transformers).

1. To utilize Core ML first, you need to convert a model from TensorFlow, PyTorch to Core ML model package format using [coremltools](https://coremltools.readme.io/docs) (or simply [utilize existing models in Core ML package format](https://github.com/likedan/Awesome-CoreML-Models) ).
2. Second, you must now use that converted package with an implementation designed for Apple Devices. Here is the [Apple XCode reference PyTorch implementation](https://github.com/apple/ml-ane-transformers).

https://machinelearning.apple.com/research/neural-engine-transformers

**Comment by qdrddr** [comment]
Work in progress on CoreML implementation for [whisper.cpp](https://github.com/ggerganov/whisper.cpp/discussions/548) you might be interested in.

**Comment by qdrddr** [comment]
https://appleinsider.com/articles/24/05/07/secret-apple-project-acdc-to-pioneer-ai-chips-for-data-centers
> Under the internal name "Project ACDC," Apple is developing [Apple Silicon](https://appleinsider.com/inside/apple-silicon) designed specifically for server farms dedicated to AI processing. The company aims to optimize AI applications within its data centers for future versions of its platforms.

**Comment by qdrddr** [comment]
2024 CoreML Updates
https://developer.apple.com/documentation/updates/coreml

WWDC 2024 sessions about CoreML:
https://developer.apple.com/wwdc24/10160
https://developer.apple.com/wwdc24/10218
https://developer.apple.com/wwdc24/10183
https://developer.apple.com/wwdc24/10161

**Comment by BrandonEchols04** [comment]
Any news on Ollama's usage of Apple Silicon's usage of the Neural Engine?

**Comment by Michael-Z-Freeman** [comment]
Until this is fixed can anyone suggest a similar project that does utilise Apple Silicon resources ?

**Comment by hg0428** [comment]
This would be a wonderful feature to have as it would allow us to do inference many times faster than what is currently possible.

**Comment by joselael** [comment]
> Until this is fixed can anyone suggest a similar project that does utilise Apple Silicon resources ?

Other than DrawThings for diffusion models, I'm not aware of any other projects that use CoreML. DrawThings is partially open source, maybe y'all can look into it to see how it's done. Otherwise, llama.cpp supports MLX, I think gregor mentioned that coreml isn't worth it yet or something. Maybe I'll make a PR for coreml when I have a long vacation someday.

**Comment by tmpbci** [comment]
not exactly like ollama but running a model with coreML https://github.com/TheMurusTeam/FreeScaler-CoreML

**Comment by johnalanwoods** [comment]
So right now Ollama can't use the ANE right?

**Comment by hg0428** [comment]
> So right now Ollama can't use the ANE right?

Correct, Ollama cannot currently use the ANE. However, they are working on this in Llama.cpp rn.

**Comment by hg0428** [comment]
https://github.com/ggerganov/llama.cpp/issues/10453

**Comment by dalisoft** [comment]
Using ANE over MLX will be more efficient, lower token/sec but uses a lot less power, less noise and lower battery usage.

There are few examples and/or repositories i'm found:
- https://github.com/smpanaro/coreml-llm-cli
- https://github.com/apple/ml-ane-transformers

**Comment by hg0428** [comment]
> Using ANE over MLX will be more efficient, lower token/sec but uses a lot less power, less noise and lower battery usage.
> 
> There are few examples and/or repositories i'm found:
> 
> * https://github.com/smpanaro/coreml-llm-cli
> * https://github.com/apple/ml-ane-transformers
> 
> And as mentioned already above
> 
> > not exactly like ollama but running a model with coreML https://github.com/TheMurusTeam/FreeScaler-CoreML

If implemented properly, it can give significantly higher token/second.
Also, the implementation being considered by Llama.cpp would not use CoreML.

**Comment by dalisoft** [comment]
@hg0428 I am newbie in LLM/AI techs so i'm only happy to see same or higher performance it is possible. Even 2x lower speed with 10x lower resource usage is good for me

**Comment by jbaenaxd** [comment]
I found this post: https://www.reddit.com/r/LocalLLaMA/comments/1iplsk1/you_can_now_run_models_on_the_neural_engine_if/

It is a new tool, it looks like it's a tool that converts the existing models to be ANE (Apple Neural Engine) compatible: https://github.com/Anemll/Anemll

They mention that it was not an easy task, apparently Apple is limiting the use of this tool.

> **Conversion Process Overview**
ANE models on iOS are limited to 1GB file size. macOS will work with ~2GB
We split models during the conversion process to avoid this limit

They already have some models that you can test, like DeepSeek R1 https://huggingface.co/anemll/anemll-DeepSeekR1-8B-ctx1024_0.1.1

**Comment by qdrddr** [comment]
Would that be of any help the [LM Studio has implemented MLX](https://github.com/lmstudio-ai/mlx-engine).

**Comment by qdrddr** [comment]
there's Apple MPS [Metal GPU flash attention in swift by the way](https://github.com/philipturner/metal-flash-attention)

**Comment by Anemll** [comment]
See https://github.com/anemll/anemll

**Comment by keefar** [comment]
There is also [https://llm.mlc.ai](url)

**Comment by qdrddr** [comment]
Metal Flash Attention PoC with Python, C, Rust, etc bindings

https://github.com/bghira/universal-metal-flash-attention

**Comment by qdrddr** [comment]
Metal flash attention in python 
https://github.com/marcogva-hub/mlx-flashattention-steel/
